index.js

```diff
@@ -16,46 +16,112 @@ class ExecutorRouter extends Executor {
 
-        executorConfig.forEach((plugin) => {
-            let ExecutorPlugin;
+        let ExecutorPlugin;
+
+        this._executors = [];
 
+        executor.forEach(plugin => {
             try {
                 // eslint-disable-next-line global-require, import/no-dynamic-require
                 ExecutorPlugin = require(`screwdriver-executor-${plugin.name}`);
+                this._executors.push(plugin);
             } catch (err) {
-                console.error(err);
+                console.error(err.message);
 
                 return;
             }
+            // Add ecosystem to executor options
+            const options = Object.assign({ ecosystem }, plugin.options);
 
-            const options = Object.assign({ ecosystem }, plugin.options); // Add ecosystem to executor options
-            const executorPlugin = new ExecutorPlugin(options);
+            this[plugin.name] = new ExecutorPlugin(options);
+        });
 
-            if (defaultPlugin && defaultPlugin === plugin.name) {
-                // set the default plugin to the desired one
-                this.defaultExecutor = executorPlugin;
-            } else if (!defaultPlugin && !this.defaultExecutor) {
-                // When given no default plugin, use the first module we instantiated
-                this.defaultExecutor = executorPlugin;
+        // executor rules chain
+        this._executorRules = [
+            {
+                name: 'annotated',
+                priority: 1,
+                check: buildConfig => {
+                    const annotations = this.parseAnnotations(buildConfig.annotations || {});
+
+                    return annotations[ANNOTATION_EXECUTOR_TYPE];
+                }
+            },
+            {
+                name: 'weighted',
+                priority: 2,
+                check: () => this.getWeightedExecutor(this._executors)
+            },
+            {
+                name: 'default',
+                priority: 3,
+                check: () => defaultPlugin || (this._executors[0] && this._executors[0].name)
             }
+        ];
+        // sort by priority and store
+        this._executorRules.sort((a, b) => a.priority - b.priority);
 
-            this[plugin.name] = executorPlugin;
-        });
-
-        if (!this.defaultExecutor) {
+        if (!this._executorRules.find(a => a.name === 'default').check()) {
             throw new Error('No default executor set.');
         }
     }
 
+    /**
+     * Returns the executor based on a random selection optimized on weightage
+     * @param {Array} executors
+     * @return {String} executor name
+     */
+    getWeightedExecutor(executors) {
+        const totalWeight = executors.reduce((prev, curr) => prev + (curr.weightage || 0), 0);
+
+        if (totalWeight === 0) {
+            return undefined;
+        }
+        const number = Math.floor(Math.random() * totalWeight);
+
+        let sum = 0;
+
+        for (let i = 0; i < executors.length; i += 1) {
```

Comment:
shouldn't u sort the `executors` first, in ascending order? Otherwise certain order in which the executor configs are loaded will be biased toward the first config